contribs/ev/src/main/java/org/matsim/contrib/ev/charging/ChargingEventSequenceCollector.java

```diff
@@ -77,6 +77,17 @@ public void handleEvent(QueuedAtChargerEvent event) {
 				"Vehicle (%s) already at a charger", event.getVehicleId());
 	}
 
+	@Override
+	public void handleEvent(QuitQueueAtChargerEvent event) {
+		Preconditions.checkState(ongoingSequences.containsKey(event.getVehicleId()),
+				"Vehicle (%s) was not at charger (%s)", event.getVehicleId(), event.getChargerId());
+		Preconditions.checkState(ongoingSequences.get(event.getVehicleId()).queuedAtCharger != null,
+				"Vehicle (%s) was not queued at charger (%s)", event.getVehicleId(), event.getChargerId());
+		Preconditions.checkState(ongoingSequences.get(event.getVehicleId()).chargingStartEvent == null,
+				"Vehicle (%s) already is already plugged", event.getVehicleId(), event.getChargerId());
+		ongoingSequences.remove(event.getVehicleId());
```

Comment:
Maybe rewrite the handler method in the following way so we call the map only once. I also added the addition of the sequence to the completed sequences (as you pointed out - otherwise we would lose this still valid sequence).
```java
		var sequence = ongoingSequences.remove(event.getVehicleId());
		Preconditions.checkState(sequence != null, "Vehicle (%s) was not at charger (%s)", event.getVehicleId(),
				event.getChargerId());
		Preconditions.checkState(sequence.queuedAtCharger != null, "Vehicle (%s) was not queued at charger (%s)",
				event.getVehicleId(), event.getChargerId());
		Preconditions.checkState(sequence.chargingStartEvent == null, "Vehicle (%s) is already plugged",
				event.getVehicleId(), event.getChargerId());
		completedSequences.add(sequence);
```

Reply:
Good point, thanks!